File: src-lang/xt/db/text/base_tree.clj
(ns xt.db.text.base-tree
  (:require [hara.lang :as l]))

(l/script :xtalk
  {:require [[xt.lang.spec-base :as xt]
             [xt.lang.common-data :as xtd]
             [xt.lang.common-tree :as xtt]
             [xt.db.text.base-check :as check]
             [xt.db.text.sql-graph :as sql-graph]
             [xt.db.text.sql-util :as sql-util]
             [xt.db.text.base-scope :as base-scope]]})

(defn.xt tree-control-array
  "creates a control array"
  {:added "4.0"}
  [control]
  (when (xtd/is-empty? control)
    (return []))
  
  (var out [])
  (var #{order-by
         order-sort
         limit
         offset} control)
  (when (xt/x:is-array? order-by)
    (xt/x:arr-push out (sql-util/ORDER-BY order-by)))
  (when (xt/x:not-nil? order-sort)
    (xt/x:arr-push out (sql-util/ORDER-SORT order-sort)))
  (when (xt/x:is-number? limit)
    (xt/x:arr-push out (sql-util/LIMIT limit)))
  (when (xt/x:is-number? offset)
    (xt/x:arr-push out (sql-util/OFFSET offset)))
  (return out))

(defn.xt tree-base
  "creates a tree base"
  {:added "4.0"}
  [schema table-name sel-query clause returning opts]
  (var tarr (base-scope/merge-queries sel-query clause))
  (var tree (xt/x:arr-assign [table-name] tarr))
  (when (xtd/not-empty? returning)
    (xt/x:arr-push tree returning))
  (return (sql-graph/select-tree schema tree opts)))

(defn.xt tree-count
  "provides a view count query"
  {:added "4.0"}
  [schema entry clause opts]
  (var #{view control} entry)
  (var #{table query} view)
  (return (-/tree-base schema table query clause (xt/x:arr-assign [{"::" "sql/count"}]
                                                                  (-/tree-control-array control))
                       opts)))

(defn.xt tree-select
  "provides a view select query"
  {:added "4.0"}
  [schema entry clause opts]
  (var #{view control} entry)
  (var #{table query} view)
  (return (-/tree-base schema table query clause (xt/x:arr-assign ["id"]
                                                                  (-/tree-control-array control))
                       opts)))

(defn.xt tree-return
  "provides a view return query"
  {:added "4.0"}
  [schema entry sel-query clause opts]
  (var #{view} entry)
  (var #{table query} view)
  (return (-/tree-base schema table sel-query clause query opts)))

(defn.xt tree-combined
  "provides a view return query"
  {:added "4.0"}
  [schema sel-entry ret-entry ret-omit clause opts]
  (var #{control} sel-entry)
  (var sel-table   (xt/x:get-path sel-entry ["view" "table"]))
  (var ret-table   (xt/x:get-path ret-entry ["view" "table"]))
  (var sel-query   (xt/x:get-path sel-entry ["view" "query"]))
  (var ret-query   (xt/x:get-path ret-entry ["view" "query"]))
  (when (xt/x:nil? sel-query)
    (:= sel-query {}))
  (when (xt/x:nil? ret-query)
    (:= ret-query {}))
  (var ret-clause  (:? (xtd/not-empty? ret-omit)
                       [{:id {:not-in [ret-omit]}}]
                       []))
  (var combined-clause (base-scope/merge-queries clause ret-clause))
  
  (return (-/tree-base schema sel-table sel-query combined-clause
                       (xt/x:arr-assign (xt/x:arr-clone ret-query)
                                        (-/tree-control-array control))
                       opts)))

(defn.xt tree-fill-input
  "fills out the tree for a given input"
  {:added "4.0"}
  [tree args input-spec drop-first]
  (var arg-map {})
  (when drop-first
    (xt/x:arr-pop-first input-spec))
  (when (== 0 (xt/x:len input-spec))
    (return tree))
  (xt/for:array [[i e] input-spec]
    (xt/x:set-key arg-map
                  (xt/x:cat "{{" (xt/x:get-key e "symbol") "}}")
                  (xt/x:get-idx args i)))
  (var out (xtt/tree-walk tree
                          (fn [x] (return x))
                          (fn [x]
                            (when (and (xt/x:is-string? x)
                                       (xt/x:has-key? arg-map x))
                              (return (xt/x:get-key arg-map x)))
                            (return x))))
  (return out))


;;
;; Plans for submission
;;

(defn.xt plan-select
  "provides a view select query"
  {:added "4.0"}
  [schema entry args opts]
  (var #{input} entry)
  (var itree  (-/tree-select schema entry {} opts))
  (var qtree  (-/tree-fill-input itree args (xt/x:arr-clone input) false))
  (return qtree))

(defn.xt plan-count
  "provides the count statement"
  {:added "4.0"}
  [schema entry args opts]
  (var #{input} entry)
  (var itree  (-/tree-count schema entry {} opts))
  (var qtree  (-/tree-fill-input itree args (xt/x:arr-clone input) false))
  (return qtree))

(defn.xt plan-return
  "provides a view return query"
  {:added "4.0"}
  [schema entry id args opts]
  (var #{input} entry)
  (var itree (-/tree-return schema entry {:id id} {} opts))
  (var qtree (-/tree-fill-input itree args (xt/x:arr-clone input) true))
  (return qtree))

(defn.xt plan-return-bulk
  "creates a bulk return statement"
  {:added "4.0"}
  [schema entry ids args opts]
  (var #{input} entry)
  (var itree  (-/tree-return schema
                             entry
                             {:id ["in" [ids]]}
                             {}
                             opts))
  (var qtree (-/tree-fill-input itree args (xt/x:arr-clone input) true))
  (return qtree))

(defn.xt plan-combined
  "provides a view combine query"
  {:added "4.0"}
  [schema sel-entry sel-args ret-entry ret-args ret-omit opts as-tree]
  (var sel-input  (xt/x:get-key sel-entry "input"))
  (var ret-input  (xt/x:get-key ret-entry "input"))
  (var itree   (-/tree-combined schema
                                sel-entry
                                ret-entry
                                ret-omit
                                []
                                opts))
  (var qtree (-/tree-fill-input itree
                                 (-> (xt/x:arr-clone  ret-args)
                                     (xt/x:arr-assign sel-args))
                                 (-> (xt/x:arr-clone  ret-input)
                                     (xt/x:arr-assign sel-input))
                                 (:? (> (xt/x:len ret-input) 0) true false)))
  (return qtree))


;;
;; PLAN VIEW
;;

(defn.xt plan-view-check
  "checks query arguments against the entry input"
  {:added "4.0"}
  [entry args drop-first]
  (var targs (xt/x:get-key entry "input"))
  (when drop-first
    (:= targs [(xt/x:unpack targs)])
    (x:arr-pop-first targs))
  (var [l-ok l-err] (check/check-args-length args targs))
  (when (not l-ok)
    (return [l-ok l-err]))
  (var [t-ok t-err] (check/check-args-type args targs))
  (when (not t-ok)
    (return [t-ok t-err]))
  (return [true nil]))

(defn.xt plan-view
  "prepares a db query from entries against a system impl"
  {:added "4.0"}
  [schema query-spec]
  (var #{table
         select-entry
         select-args
         return-entry
         return-args
         return-omit
         return-count
         return-id
         return-bulk} query-spec)
  (cond (and (xt/x:not-nil? select-entry)
             (xt/x:not-nil? return-entry))
        (do (var [s-ok s-err] (-/plan-view-check select-entry select-args false))
            (when (not s-ok)
              (return [s-ok s-err]))
            (var [r-ok r-err] (-/plan-view-check return-entry return-args true))
            (when (not r-ok)
              (return [r-ok r-err]))
            (return [true (-/plan-combined
                           schema
                           select-entry
                           select-args
                           return-entry
                           return-args
                           return-omit
                           {}
                           false)]))
        
        (xt/x:not-nil? select-entry)
        (do (var [s-ok s-err] (-/plan-view-check select-entry select-args false))
            (when (not s-ok)
              (return [s-ok s-err]))
            (return [true (:? return-count
                              (-/plan-count
                               schema
                               select-entry
                               select-args
                               {})
                              (-/plan-select
                               schema
                               select-entry
                               select-args
                               {}))]))

        (xt/x:not-nil? return-id)
        (do (var rargs [return-id (xt/x:unpack return-args)])
            (var [r-ok r-err] (-/plan-view-check return-entry rargs false))
            (when (not r-ok)
              (return [r-ok r-err]))
            (return [true (-/plan-return
                           schema
                           return-entry
                           return-id
                           return-args
                           {})]))

        (xt/x:not-nil? return-bulk)
        (do (var [r-ok r-err] (-/plan-view-check return-entry return-args true))
            (when (not r-ok)
              (return [r-ok r-err]))
            (return [true (-/plan-return-bulk
                           schema
                           return-entry
                           return-bulk
                           return-args
                           {})]))

        :else
        (return [true nil])))

File: packages-gen/js/libs/xt-db/text/base-tree.js
const xtt = require("@xtalk/lang/common-tree.js")

const xtd = require("@xtalk/lang/common-data.js")

const base_scope = require("@xtalk/db/text/base-scope.js")

const check = require("@xtalk/db/text/base-check.js")

const sql_util = require("@xtalk/db/text/sql-util.js")

const sql_graph = require("@xtalk/db/text/sql-graph.js")

function tree_control_array(control){
  if(xtd.is_emptyp(control)){
    return [];
  }
  let out = [];
  let {limit,offset,order_by,order_sort} = control;
  if(Array.isArray(order_by)){
    out.push(sql_util.ORDER_BY(order_by));
  }
  if(null != order_sort){
    out.push(sql_util.ORDER_SORT(order_sort));
  }
  if("number" == (typeof limit)){
    out.push(sql_util.LIMIT(limit));
  }
  if("number" == (typeof offset)){
    out.push(sql_util.OFFSET(offset));
  }
  return out;
}

function tree_base(schema,table_name,sel_query,clause,returning,opts){
  let tarr = base_scope.merge_queries(sel_query,clause);
  let tree = xtd.arr_assign([table_name],tarr);
  if(xtd.not_emptyp(returning)){
    tree.push(returning);
  }
  return sql_graph.select_tree(schema,tree,opts);
}

function tree_count(schema,entry,clause,opts){
  let {control,view} = entry;
  let {query,table} = view;
  return tree_base(
    schema,
    table,
    query,
    clause,
    xtd.arr_assign([{"::":"sql/count"}],tree_control_array(control)),
    opts
  );
}

function tree_select(schema,entry,clause,opts){
  let {control,view} = entry;
  let {query,table} = view;
  return tree_base(
    schema,
    table,
    query,
    clause,
    xtd.arr_assign(["id"],tree_control_array(control)),
    opts
  );
}

function tree_return(schema,entry,sel_query,clause,opts){
  let {view} = entry;
  let {query,table} = view;
  return tree_base(schema,table,sel_query,clause,query,opts);
}

function tree_combined(schema,sel_entry,ret_entry,ret_omit,clause,opts){
  let {control} = sel_entry;
  let sel_table = sel_entry["view"]["table"];
  let ret_table = ret_entry["view"]["table"];
  let sel_query = sel_entry["view"]["query"];
  let ret_query = ret_entry["view"]["query"];
  if(null == sel_query){
    sel_query = {};
  }
  if(null == ret_query){
    ret_query = {};
  }
  let ret_clause = xtd.not_emptyp(ret_omit) ? [{"id":{"not_in":[ret_omit]}}] : [];
  let combined_clause = base_scope.merge_queries(clause,ret_clause);
  return tree_base(
    schema,
    sel_table,
    sel_query,
    combined_clause,
    xtd.arr_assign(ret_query.slice(),tree_control_array(control)),
    opts
  );
}

function tree_fill_input(tree,args,input_spec,drop_first){
  let arg_map = {};
  if(drop_first){
    input_spec.shift();
  }
  if(0 == input_spec.length){
    return tree;
  }
  for(let i = 0; i < input_spec.length; ++i){
    let e = input_spec[i];
    arg_map["{{" + e["symbol"] + "}}"] = args[i];
  };
  let out = xtt.tree_walk(tree,function (x){
    return x;
  },function (x){
    if(("string" == (typeof x)) && (null != arg_map[x])){
      return arg_map[x];
    }
    return x;
  });
  return out;
}

function plan_select(schema,entry,args,opts){
  let {input} = entry;
  let itree = tree_select(schema,entry,{},opts);
  let qtree = tree_fill_input(itree,args,input.slice(),false);
  return qtree;
}

function plan_count(schema,entry,args,opts){
  let {input} = entry;
  let itree = tree_count(schema,entry,{},opts);
  let qtree = tree_fill_input(itree,args,input.slice(),false);
  return qtree;
}

function plan_return(schema,entry,id,args,opts){
  let {input} = entry;
  let itree = tree_return(schema,entry,{"id":id},{},opts);
  let qtree = tree_fill_input(itree,args,input.slice(),true);
  return qtree;
}

function plan_return_bulk(schema,entry,ids,args,opts){
  let {input} = entry;
  let itree = tree_return(schema,entry,{"id":["in",[ids]]},{},opts);
  let qtree = tree_fill_input(itree,args,input.slice(),true);
  return qtree;
}

function plan_combined(schema,sel_entry,sel_args,ret_entry,ret_args,ret_omit,opts,as_tree){
  let sel_input = sel_entry["input"];
  let ret_input = ret_entry["input"];
  let itree = tree_combined(schema,sel_entry,ret_entry,ret_omit,[],opts);
  let qtree = tree_fill_input(
    itree,
    xtd.arr_assign(ret_args.slice(),sel_args),
    xtd.arr_assign(ret_input.slice(),sel_input),
    (ret_input.length > 0) ? true : false
  );
  return qtree;
}

function plan_view_check(entry,args,drop_first){
  let targs = entry["input"];
  if(drop_first){
    targs = [...targs];
    targs.shift();
  }
  let [l_ok,l_err] = check.check_args_length(args,targs);
  if(!l_ok){
    return [l_ok,l_err];
  }
  let [t_ok,t_err] = check.check_args_type(args,targs);
  if(!t_ok){
    return [t_ok,t_err];
  }
  return [true,null];
}

function plan_view(schema,query_spec){
  let {return_args,return_bulk,return_count,return_entry,return_id,return_omit,select_args,select_entry,table} = query_spec;
  if((null != select_entry) && (null != return_entry)){
    let [s_ok,s_err] = plan_view_check(select_entry,select_args,false);
    if(!s_ok){
      return [s_ok,s_err];
    }
    let [r_ok,r_err] = plan_view_check(return_entry,return_args,true);
    if(!r_ok){
      return [r_ok,r_err];
    }
    return [
      true,
      plan_combined(
          schema,
          select_entry,
          select_args,
          return_entry,
          return_args,
          return_omit,
          {},
          false
        )
    ];
  }
  else if(null != select_entry){
    let [s_ok,s_err] = plan_view_check(select_entry,select_args,false);
    if(!s_ok){
      return [s_ok,s_err];
    }
    return [
      true,
      return_count ? plan_count(schema,select_entry,select_args,{}) : plan_select(schema,select_entry,select_args,{})
    ];
  }
  else if(null != return_id){
    let rargs = [return_id,...return_args];
    let [r_ok,r_err] = plan_view_check(return_entry,rargs,false);
    if(!r_ok){
      return [r_ok,r_err];
    }
    return [
      true,
      plan_return(schema,return_entry,return_id,return_args,{})
    ];
  }
  else if(null != return_bulk){
    let [r_ok,r_err] = plan_view_check(return_entry,return_args,true);
    if(!r_ok){
      return [r_ok,r_err];
    }
    return [
      true,
      plan_return_bulk(schema,return_entry,return_bulk,return_args,{})
    ];
  }
  else{
    return [true,null];
  }
}

module.exports = {
  ["tree_control_array"]:tree_control_array,
  ["tree_base"]:tree_base,
  ["tree_count"]:tree_count,
  ["tree_select"]:tree_select,
  ["tree_return"]:tree_return,
  ["tree_combined"]:tree_combined,
  ["tree_fill_input"]:tree_fill_input,
  ["plan_select"]:plan_select,
  ["plan_count"]:plan_count,
  ["plan_return"]:plan_return,
  ["plan_return_bulk"]:plan_return_bulk,
  ["plan_combined"]:plan_combined,
  ["plan_view_check"]:plan_view_check,
  ["plan_view"]:plan_view
}